src/Smalot/PdfParser/Font.php

```diff
@@ -381,86 +393,182 @@ public function decodeText(array $commands): string
+    }
+
+    private function getInitializedEncodingByPdfObject(PDFObject $PDFObject): Encoding
+    {
+        if (!$this->initializedEncodingByPdfObject) {
+            $this->initializedEncodingByPdfObject = $this->createInitializedEncodingByPdfObject($PDFObject);
+        }
+
+        return $this->initializedEncodingByPdfObject;
+    }
+
+    /**
+     * Decode content when Encoding is instance of Encoding.
+     */
+    private function decodeContentByEncodingEncoding(string $text, Encoding $encoding): string
+    {
+        $result = '';
+        $length = \strlen($text);
+
+        for ($i = 0; $i < $length; ++$i) {
+            $dec_av = hexdec(bin2hex($text[$i]));
+            $dec_ap = $encoding->translateChar($dec_av);
+            $result .= self::uchr($dec_ap ?? $dec_av);
+        }
+
+        return $result;
+    }
+
+    /**
+     * Decode content when Encoding is instance of Element.
+     */
+    private function decodeContentByEncodingElement(string $text, Element $encoding): ?string
+    {
+        $pdfEncodingName = $encoding->getContent();
+
+        // mb_convert_encoding does not support MacRoman/macintosh,
+        // so we use iconv() here
+        $iconvEncodingName = $this->getIconvEncodingNameOrNullByPdfEncodingName($pdfEncodingName);
+
+        return $iconvEncodingName ? iconv($iconvEncodingName, 'UTF-8', $text) : null;
+    }
+
+    /**
+     * Convert PDF encoding name to iconv-known encoding name.
+     */
+    private function getIconvEncodingNameOrNullByPdfEncodingName(string $pdfEncodingName): ?string
+    {
+        $pdfToIconvEncodingNameMap = [
+            'StandardEncoding' => 'ISO-8859-1',
+            'MacRomanEncoding' => 'MACINTOSH',
+            'WinAnsiEncoding' => 'CP1252',
+        ];
+
+        return \array_key_exists($pdfEncodingName, $pdfToIconvEncodingNameMap)
+            ? $pdfToIconvEncodingNameMap[$pdfEncodingName]
+            : null;
+    }
+
+    private function decodeContentByAutodetectIfNecessary($text)
+    {
+        if (mb_check_encoding($text, 'UTF-8')) {
+            return $text;
+        }
+
+        return mb_convert_encoding($text, 'UTF-8', 'Windows-1252');
+        //todo: Why exactly `Windows-1252` used?
+    }
+
+    /**
+     * Create Encoding instance by PDFObject instance and init it.
+     */
+    private function createInitializedEncodingByPdfObject(PDFObject $PDFObject): Encoding
+    {
+        $encoding = $this->createEncodingByPdfObject($PDFObject);
+        $encoding->init();
+
+        return $encoding;
+    }
+
+    /**
+     * Create Encoding instance by PDFObject instance (without init).
+     */
+    private function createEncodingByPdfObject(PDFObject $PDFObject): Encoding
+    {
+        $document = $PDFObject->getDocument();
+        $header = $PDFObject->getHeader();
+        $content = $PDFObject->getContent();
+        $config = $PDFObject->getConfig();
+
+        return new Encoding($document, $header, $content, $config);
+    }
```

Comment:
That is a bit over engineered in my opinion. `createEncodingByPdfObject` and `createInitializedEncodingByPdfObject` are only used once and are basically shortcuts. What is the benefit of having them? Why not init related objects at the place where they are need?

Reply:
> That is a bit over engineered in my opinion ... Why not init related objects at the place where they are needed?

I don't think so. This methods just implements "details" in "main" algorithm. This reduces cognitive load when read "main" algorithm logic.

If you implement difficult algorithm in just one method it could grow up to thousands of lines and it would be too difficult or even impossible to understand what method realy do, because you must scrool too long and during reading code sort in your mind which lines are really important for implemented algorithm and which is just noise with unimportant details implementation.

Imho, it's much better when "main" method has just 10 lines and 10 nested methods with 10 lines for each, then one method with has 100 lines of code.

Look at Font::decodeContent() method before (84 lines of code with many deep nested if/else/foreach):
https://github.com/smalot/pdfparser/blob/e056671d50b31d6320a26aa8fbfafd2c02e6bff1/src/Smalot/PdfParser/Font.php#L381-L465

and after this PR (15 lines of code with just 2 simple if-return statements):

https://github.com/likemusic/pdfparser/blob/21c4797009c23979005fc6fb1ee358c2abd8dfb8/src/Smalot/PdfParser/Font.php#L393-L408

I think it's obvious that extracting some line groups to separated methods made it much easier to understand the logic of the method.


Splitting to createEncodingByPdfObject() and createInitializedEncodingByPdfObject() made to show that created object could has different states (not initialized and initialized).


By the way, I think that using initialized and not initialized object is not good idea, because in this case after geting object you must always check his state and init it if requred. It's more easy to use objects when they ready for usage after just creating it by constructor. So I think it would be much better if initialized and not initialized objects would be presented by different type groups, where initialized inherits or aggregate not initialized type.

Reply:
I agree with the fact that this library has code parts which are not good structured and have smell.

But this fact doesn't support your suggested change. When I take the code of your functions and combine it, it boils down to:

```php
$encoding = new Encoding(
    $PDFObject->getDocument(), 
    $PDFObject->getHeader(), 
    $PDFObject->getContent(), 
    $PDFObject->getConfig()
);
$encoding->init();
```

Its size is small and everyone get its meaning quickly, in my opinion. If in doubt a comment might help.

---

If you still don't agree with me I can accept the fact that you want to keep these 2 functions. Your underlying approach is to apply a better structure to the library and improve code readability which I welcome and support. I don't want to risk loosing your commitment over a two functions :)

Reply:
> When I take the code of your functions and combine it ...

You mix in one function two different operations: instantiating and configuring.
Imho, it is always better to have a low-level function that does just only one thing and hi-level functions that use the low-level function.

In this case, configuring is just calling `init()`. Therefore I don't extract it to another function)) But often configuring objects after creation is not so simple.

I think my version is better because it could help for the next factoring iteration (when $encoding would be private class member instantiated always like Encoding in Font constructor or when objects would not require call init() before using them).

Reply:
> I don't want to risk loosing your commitment over two functions :)

Thank you! I hope it will inspire me to fix some another encoding issues :)

And I understand that introducing a new function just for abstract one-line object configuration from creation seems like over-engineering, but I think for the long term my variant is better.

I feel that soon I'll understand how pdf-files encodings must be decoded and how it does most popular readers/viewers))

Reply:
👍🏿 Resolved for me, see my comments at https://github.com/smalot/pdfparser/pull/500#discussion_r780224706